beluga/include/beluga/sensor/likelihood_field_model3.hpp

```diff
@@ -0,0 +1,168 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef BELUGA_SENSOR_LIKELIHOOD_FIELD_MODEL3_HPP
+#define BELUGA_SENSOR_LIKELIHOOD_FIELD_MODEL3_HPP
+
+#include <algorithm>
+#include <cmath>
+#include <random>
+#include <vector>
+
+#include <openvdb/openvdb.h>
+
+#include <Eigen/Core>
+
+#include <range/v3/range/conversion.hpp>
+#include <range/v3/view/all.hpp>
+#include <range/v3/view/transform.hpp>
+#include <sophus/se3.hpp>
+#include <sophus/so3.hpp>
+
+/**
+ * \file
+ * \brief Implementation of a likelihood field sensor model for 3D Lidars.
+ */
+
+namespace beluga {
+
+/// Parameters used to construct a LikelihoodFieldModel3 instance.
+/**
+ * See Probabilistic Robotics \cite thrun2005probabilistic Chapter 6.4, particularly Table 6.3.
+ */
+struct LikelihoodFieldModel3Param {
+  /// Maximum distance to obstacle.
+  /**
+   * When creating a distance map, if the distance to an obstacle is higher than the value specified here,
+   * then this value will be used.
+   */
+  double max_obstacle_distance = 100.0;
+  /// Maximum range of a laser ray.
+  double max_laser_distance = 2.0;
+  /// Weight used to combine the probability of hitting an obstacle.
+  double z_hit = 0.5;
+  /// Weight used to combine the probability of random noise in perception.
+  double z_random = 0.5;
+  /// Standard deviation of a gaussian centered arounds obstacles.
+  /**
+   * Used to calculate the probability of the obstacle being hit.
+   */
+  double sigma_hit = 0.2;
+};
+
+/// Likelihood field sensor model for range finders.
+/**
+ * This model relies on a pre-computed likelihood map of the environment.
+ * It is less computationally intensive than the beluga::BeamSensorModel
+ * because no ray-tracing is required, and it can also provide better
+ * performance in environments with non-smooth occupation maps. See
+ * Probabilistic Robotics \cite thrun2005probabilistic, Chapter 6.4,
+ * for further reference.
+ *
+ * \note This class satisfies \ref SensorModelPage.
+ *
+ * \tparam OpenVDB grid type.
+ */
+template <typename GridT, typename PointCloud>
+class LikelihoodFieldModel3 {
+ public:
+  /// State type of a particle.
+  using state_type = Sophus::SE3d;
+
+  /// Weight type of the particle.
+  using weight_type = double;
+
+  /// Measurement type given by the interface.
+  using measurement_type = PointCloud;
+
+  /// Map representation type.
+  using map_type = GridT;
+
+  /// Parameter type that the constructor uses to configure the likelihood field model.
+  using param_type = LikelihoodFieldModel3Param;
+
+  /// Constructs a Likelihood3DFieldModel instance.
```

Comment:
There are other instances that need replacement.
```suggestion
  /// Constructs a LikelihoodFieldModel3 instance.
```

Reply:
Changed in 2aabb60e1d02fcf2464ebca96f3cdb47be230d3a